frappe/erpnext
# refactor!: Remove HR and Payroll modules from ERPNext

## New App

HR & Payroll modules are now going to be a part of a new app https://github.com/frappe/hrms.
This app still depends on ERPNext.

## Why not standalone?

Unlike other modules that were separated from the ERPNext core, HR and Payroll have a high dependency on the core. Some doctypes from HR are also widely used in the ERPNext core.

## Separation Details

### Generic

- The following doctypes are used in ERPNext and hence they are moved from **HR** module to **Setup** module in ERPNext:

	- **Employee** 😢: used in Telephony, Stock, Quality, Projects, Manufacturing, Loans, Buying, Assets, Accounts, Dependent Apps: Healthcare
	- **Holiday List**: CRM (Appointment Booking), Projects, Healthcare, Support (SLA), Education (Student's Holiday List)
	- **Branch**: Selling (SMS Center), used widely as an accounting dimension
	- **Department**: Assets, Projects, SMS Center, used widely as an accounting dimension
	- **Designation**: Lead, Healthcare Practitioner, used in Employee master which will now be in core
	- **Employee Group**: Telephony, Communication
	- **Child Tables for above doctypes**: Employee Group Table (Employee Group), Employee Education (Employee), Employee External Work History (Employee), Employee Internal Work History (Employee), Holiday (Holiday List)
	- **Driver, Vehicle, Driving License Category (child table)**: Delivery Trip
- Fields in the above masters that are HR-specific have been removed and would be added as Custom Fields on the HRMS app setup: https://github.com/frappe/hrms/commit/4d0aa1fafb04a3c0915688ee01bc95bdb9763f31, https://github.com/frappe/hrms/commit/6a1caa7ef04e1aa2186de7c585c327b829eb86e4
	<details>
		<summary>Fields</summary>
		
		Department
		  1. payroll_cost_center
		  2. leave_block_list
		  3. approvers
		  4. leave_approvers
  		  5. expense_approvers
  		  6. shift_request_approver

		Employee
			1. employment_type
			2. job_applicant
			3. grade
			4. default_shift
			5. health_insurance_section
			6. health_insurance_provider
			7. health_insurance_no
	</details> 
- References to HR/Payroll doctypes in other form dashboards has been removed. Its added in the HRMS app using `override_doctype_dashboards` hook: https://github.com/frappe/hrms/commit/6f9ef74c42da73aeee420e66c503087efac71db4
- Removed HR reports from other modules: 	https://github.com/frappe/hrms/commit/da9f190b5ad00e955edd983cc52f4d45f74f9551	
	- Project Profitability report: moved from Projects
	- Employee Hours Utilization Based on Timesheet: moved from Projects
	- Unpaid Expense Claims: moved from Accounts
- **HR** field removed from the "Terms and Conditions" doctype. HR specific Email Templates and Notifications moved. User Type and Permission setup moved: https://github.com/frappe/hrms/commit/c958762a59b0392f1fd9093b5cdf846f77d1dd83
- Removed HR and Payroll links from ERPNext Workspaces

### Regional

Moved HR/Payroll Code from the **Regional** module:
- HRA Tax Exemption Calculation 
- Regional Setups
	- India: Region-specific custom fields, salary components, and gratuity rules
	- UAE: Gratuity Rules
- Reports: Professional Tax Deduction and Provident Fund Deductions

https://github.com/frappe/hrms/commit/7f6a1942c3f9107376b739394ec94e68503f3319

### Loan Management

- **Repay from Salary** field removed from Loan, Loan Repayment. Will be added as a custom field on HRMS setup
- Moved Salary Slip Loan child table to HRMS app
- Fixed references to Salary Slip Loan and `repay_from_salary` in queries

https://github.com/frappe/hrms/commit/428cb2dd75668ce657e15d1b0a6b898217faa34d, https://github.com/frappe/hrms/commit/00bbdae6cf97518f8b19db6edc879730a466f9b7

### Accounting

Removed HR/Payroll code from Payment Entry and Journal Entry: 
- Expense Claim
- Employee Advance
- Full and Final Statement
- Gratuity

ERPNext: https://github.com/frappe/erpnext/commit/1ff0e4519f5f5d03348d31e99a980c0c4b7701b8, https://github.com/frappe/erpnext/commit/0f83e7af56176efd7be4870979318fa414160a84
HRMS: https://github.com/frappe/hrms/commit/69b951901f09595d99ece4ffff3d24e2546fa85f, https://github.com/frappe/hrms/commit/a9b2ba1d68041378cd0607e01298ea64d99ecfb6

### Projects

- Removed **Create Salary Slip** workflow from Timesheet. https://github.com/frappe/hrms/commit/5956f2d0d06311d4dba2ccaf13bcae3b7705cc3c
- Removed `total_expense_claim` field and Expense Claim calculation from Project and Task: https://github.com/frappe/hrms/commit/68ee6e653cced9aa6010cb8930270ab104e89543

### Delivery Trip

- Removed Expense Claim creation flow from Delivery Trip: https://github.com/frappe/hrms/commit/68ee6e653cced9aa6010cb8930270ab104e89543

### Testing Details

<details>
	<summary>To Test</summary>

1. ESS User Role is created and is able to access required doctypes
2. The following reports were moved from other modules to HR. Check if they are working fine with some data:
	- Project Profitability report: moved from Projects
	- Employee Hours Utilization Based on Timesheet: moved from Projects
	- Unpaid Expense Claims: moved from Accounts
	- Professional Tax Deduction
	- Provident Fund Deductions
3. Following dashboards are overridden with the HRMS app. Check if all links show up correctly and navigation is working:
	- Employee
	- Holiday List
	- Project
	- Timesheet
4. Check if HRA Tax Exemption Calculation is working fine for an Indian company
5. Test **Repay from Salary** flow in Loans.
6. Payment Entry and Journal Entry against:
	- Employee Advance. Check if the status is changing properly on Payment submission/cancellation.
	- Expense Claim. Check if the status is changing to Paid/Unpaid (on cancellation) properly.
7. Payment against Gratuity
8. **Create Salary Slip** workflow against Timesheet
9. Link Project to Expense Claim and test if the **Total Expense Claim** field gets updated properly whenever Expense Claim is created against a Project and Task.
10. Expense Claim creation against Delivery Trip
11. Bank Reconciliation for Expense Claim
12. Check if Payroll Entry is working fine (start to end)

</details>

**ToDo**

- [x] Expense Claim in Bank Reconciliation Tool and Bank Transaction
- [x] HR and Payroll Settings in Company master
- [x] Vehicle, Driver masters from HR being used in Delivery Trip. Move to Setup module
- [x] Patches (https://github.com/frappe/hrms/pull/3)
- [x] Review roles and permissions (keeping them as is since HR is still dependent on ERPNext)
